packages/components/src/post-locked-modal/index.js

```diff
@@ -0,0 +1,120 @@
+import {
+	PostPreviewButton,
+} from '@wordpress/editor';
+import { withSelect } from '@wordpress/data';
+
+import './style.scss';
+import { addQueryArgs } from '@wordpress/url';
+import { getPostEditURL } from '../../../../edit-post/components/browser-url';
+import { getWPAdminURL } from '../../../editor/src/utils/url';
+
+class PostLockedModal extends Component {
+	constructor() {
+		super( ...arguments );
+		this.state = {
+			isOpen: true,
+		};
+
+		this.takeOverPost = this.takeOverPost.bind( this );
+		const { lockDetails } = this.props;
+
+		if ( lockDetails && lockDetails.text ) {
+			this.modalText = lockDetails.text;
+			this.avatar = lockDetails.avatar_src;
+			this.takeover = true;
+		} else {
+			const { user } = this.props;
+			const displayName = ( user && user.data ) ? user.data.display_name : __( 'Another user' );
+			this.modalText = sprintf( __( '%s is already editing this post. Do you want to take over?' ), displayName );
+			this.avatar = user.data.avatar_src;
+		}
+	}
+
+	takeOverPost() {
+		const { lockNonce, id } = this.props();
+		const unlockUrl = addQueryArgs( getPostEditURL(), {
+			'get-post-lock': '1',
+			_wpnonce: lockNonce,
+			lockKey: true,
+			post: id,
+		} );
+		document.location = unlockUrl;
+	}
+
+	allPosts() {
+		document.location = getWPAdminURL( 'edit.php' );
+	}
+
+	render() {
+		return (
+			<Fragment>
+				{
+					this.state.isOpen ?
+						<Modal
+							title={ this.takeover ? __( 'Post taken over' ) : __( 'Post locked' ) }
+							onRequestClose={ this.closeModal }
+							focusOnMount={ true }
+							shouldCloseOnClickOutside={ false }
+							shouldCloseOnEsc={ false }
+							isDismissable={ false }
+							className="post-locked-modal"
+						>
+							{
+								this.avatar &&
+									<img
+										src={ this.avatar }
+										alt={ __( 'Avatar' ) }
+										className="components-modal__image"
+									/>
+							}
+							<span>
+								<div>
+									{ this.modalText }
+								</div>
+								{
+									this.takeover ?
+										<p><a
+											href={ getWPAdminURL( 'edit.php' ) }
+										>
+											{ __( 'View all posts' ) }
+										</a></p> :
+										<button
+											className={ 'button' }
+											onClick={ this.allPosts }
+										>
+											{ __( 'All Posts' ) }
+										</button>
+								}
+								{ ! this.takeover &&
+									<span>
+										<PostPreviewButton />
+										<button
+											className="button button-primary"
+											onClick={ this.takeOverPost }
+										>
+											{ __( 'Take Over' ) }
+										</button>
+									</span>
+								}
+							</span>
+						</Modal> :
+						null
+				}
+			</Fragment>
+		);
+	}
+}
+withSelect( ( select ) => {
+	return {
+		lockDetails: select( 'core/edit-post' ).getLockDetails(),
+		user: select( 'core/edit-post' ).getPostLockUser(),
+		id: select( 'core/edit-post' ).getCurrentPost(),
+		lockNonce: select( 'core/edit-post' ).getEditorSettings(),
```

Comment:
I think these should be `core/editor` instead of `core/edit-post` right?

Reply:
ah, right :) thanks.

Reply:
I think it also needs to be exported 😉

Reply:
yea, that was incomplete, i reworked in https://github.com/WordPress/gutenberg/pull/4217/commits/a7987b8c15ecbe1414db63362f9f16e8b3bd8a57